ground/gcs/src/plugins/uavtalk/uavtalk.cpp

```diff
@@ -191,244 +182,128 @@ bool UAVTalk::objectTransaction(UAVObject *obj, quint8 type, bool allInstances)
-            rxState = STATE_SYNC;
-            UAVTALK_QXTLOG_DEBUG("UAVTalk: Type->Sync");
-            break;
-        }
-
-        rxType = rxbyte;
-
-        packetSize = 0;
-
-        rxState = STATE_SIZE;
-        UAVTALK_QXTLOG_DEBUG("UAVTalk: Type->Size");
-        rxCount = 0;
-        break;
+    if (bytesAvail < sizeof(UAVTalkHeader)) {
+        return false;
+    }
 
-    case STATE_SIZE:
+    UAVTalkHeader *hdr = (UAVTalkHeader *) (rxBuffer + startOffset);
 
-        // Update CRC
-        rxCS = updateCRC(rxCS, rxbyte);
+    /* Basic framing checks.  If these fail, skip forward one byte and retry
+     * to capture stream sync.
+     */
+    if (hdr->sync != SYNC_VAL) {
+        startOffset++;
+        stats.rxErrors++;
 
-        if (rxCount == 0) {
-            packetSize += rxbyte;
-            rxCount++;
-            UAVTALK_QXTLOG_DEBUG("UAVTalk: Size->Size");
-            break;
-        }
+        return true;
+    }
 
-        packetSize += (quint32)rxbyte << 8;
+    if ((hdr->type & VER_MASK) != TYPE_VER) {
+        startOffset++;
+        stats.rxErrors++;
 
-        if (packetSize < MIN_HEADER_LENGTH
-            || packetSize > MAX_HEADER_LENGTH + MAX_PAYLOAD_LENGTH) { // incorrect packet size
-            rxState = STATE_SYNC;
-            UAVTALK_QXTLOG_DEBUG("UAVTalk: Size->Sync");
-            break;
-        }
+        return true;
+    }
 
-        rxCount = 0;
-        rxState = STATE_OBJID;
-        UAVTALK_QXTLOG_DEBUG("UAVTalk: Size->ObjID");
-        break;
+    if (hdr->size < sizeof(UAVTalkHeader)) {
+        startOffset++;
+        stats.rxErrors++;
 
-    case STATE_OBJID:
+        return true;
+    }
 
-        // Update CRC
-        rxCS = updateCRC(rxCS, rxbyte);
+    /* OK, let's ensure we have enough bytes for the whole frame. 
+     * Size doesn't include CRC, so add one.
+     */
 
-        rxTmpBuffer[rxCount++] = rxbyte;
-        if (rxCount < 4) {
-            UAVTALK_QXTLOG_DEBUG("UAVTalk: ObjID->ObjID");
-            break;
-        }
+    if ((hdr->size + 1u) > bytesAvail) {
+        return false;
+    }
 
-        // Search for object, if not found reset state machine
-        rxObjId = (qint32)qFromLittleEndian<quint32>(rxTmpBuffer);
-        {
-            UAVObject *rxObj = objMngr->getObject(rxObjId);
-            if (rxObj == NULL && rxType != TYPE_OBJ_REQ) {
-                stats.rxErrors++;
-                rxState = STATE_SYNC;
-                UAVTALK_QXTLOG_DEBUG("UAVTalk: ObjID->Sync (badtype)");
-                break;
-            } else if (rxObj == NULL) {
-                // This is a non-existing object, just skip to checksum
-                // and we'll send a NACK next.
-                rxState = STATE_CS;
-                UAVTALK_QXTLOG_DEBUG("UAVTalk: ObjID->CSum (no obj)");
-                rxInstId = 0;
-                rxCount = 0;
-                break;
-            }
+    quint8 ourCrc = updateCRC(0, rxBuffer + startOffset, hdr->size);
```

Comment:
Could-- but that's both slightly move expensive and relies on a property of CRCs that not all may know.